scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala

```diff
@@ -149,11 +150,19 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
     *
     * @return A random Float, and the next Randomizer to use.
     */
-  def nextFloat: (Float, Randomizer) = { // Uses same algorithm as ScalaCheck for this one
-    val (s, rs) = chooseInt(0, 1)
-    val (e, re) = chooseInt(0, 0xfe)
-    val (m, rm) = chooseInt(0, 0x7fffff)
-    (java.lang.Float.intBitsToFloat((s << 31) | (e << 23) | m), rm)
+  def nextFloat: (Float, Randomizer) = {
+    // The space of possible Floats
+    // 2 to the power of 32, which is covered by Int.MinValue to Int.MaxValue
+    // - (2 to the power of 23 (23 is the number of mantissa bits))
+    // + (3 for the two infinities and one NaN)
+
+    // scala> Int.MaxValue - math.pow(2, 23).toInt + 2
+    // res13: Int = 2139095042
+    val (x, r) = chooseInt(Int.MinValue, 2139095042) 
+
+    // Pick one lucky number to play the lotto with:
+    if (x == 999) (Float.NaN, r)
```

Comment:
While I can understand that, I think you're introducing a real problem here: you've built a better mousetrap, and expecting folks to only use it in one room is probably a mistake.

That is, what you've got here, intentionally, is a much better replacement for `java.util.Random`.  Specifically, you have a replacement that is designed to work hand-in-glove with Scalactic and its types.

That being the case, I would expect anybody *using* Scalactic who is using randomness for *anything* to wind up using this class.  Saying "It is not intended to be used in production" is probably futile: folks sometimes *do* use randomness in production code, and you've got a rich randomness library, so I expect people to use it.

I suppose what I'm arguing here is that Randomizer is in the wrong place.  It probably belongs in Scalactic, not in ScalaTest, and should be considered with that in mind.  You've built a more general tool than I think you may have intended, and I suspect folks will use it as such...

Reply:
Well, gnerating random numbers for testing is a different use case than the many different use cases for random numbers someone might want in production. These aren't secure. In production you may or may not want security. They favor touching the whole space of possibility over raw speed. The replacement nextFloat/DoubleBetween0And1 algo is probably slower than the one used in java.util.Random, but touches all the points in the space of possibilities, whereas the one in java.util.Random does not. This Randomizer generates undesirable values like infinities and NaNs, which are useful in testing, but not in production. This Randomizer requires that you pass the thing through each time in a functional way so that users can get repeatable test results. That's nicely functional, so maybe people want that in production too, but that's different from java.util.Random.

If you feel the urge to make random number generator that knows about Scalactic anyvals that's a different problem. I have never heard of that need, so I don't know anything about what people might want. It seems plausible, but that's a different design for a different day.

Reply:
Well, my final word on the subject is, at least make the ScalaDoc more explicit about this.  It's not a topic I was thinking about when I was documenting `nextFloat()` and `nextDouble()` (didn't even occur to me), but if those can return NaNs and infinities, it should be crystal-clear that it's going to do so, because otherwise I think that many (I suspect most) folks won't even consider that possibility.  It's a driving reason why they may want to seek out other functions instead...

Reply:
For sure. I noticed a few spots where we need to add more information, such as saying the Between methods are inclusive of 0 but exclusive of 1. I'll do a pass over the docs before we release.